### ISSUE TITLE
Automate: RBAC and Authentication

### ISSUE THREAD

**Issue by jefkel** [BCDevOps/OpenShift4-RollOut#181]
Configure authentication and platform service RBAC setup.

- [x] Set up GitHub authentication, delete default kubeadmin
- [x] Cluster Admins, Cluster Viewers
- [x] BCDevExch Service account and "oc --as" permission setup
- [x] Restricted project/namespace creation

**Comment by j-pye** [comment]
I've been reading through the Github and IdentityProvider setup for OCP4. This might be a case where a throw away cluster would be useful for testing from scratch @filiy 
Just so i can start from nothing and test setting up auth with Github.

I was hoping the GIthub setup could be automated via the Github API but it looks like that functionality is being removed from their API in favor of the github web interface.(boo!) So it looks like part of this will be semi-manual on the github side.

Think I have the last 3 bullets sorted. Going to go with a cron job as suggested but if there is any desire for this to be an operator then I can make that work as well. :)

Also, @sbarre-esit @jefkel  are we planning to keep RBAC in https://github.com/bcgov-c/devops-platform-operations-docs/tree/master/rbac ?

**Comment by StevenBarre** [comment]
Pretty sure we can handle this with a template and some vaulted vars. See my [other docs](https://github.com/bcgov-c/advsol-docs/blob/master/OCP-4-Install.md#configure-github-auth) for the manual set of steps.

```yaml
apiVersion: config.openshift.io/v1
kind: OAuth
metadata:
  name: cluster
spec:
  identityProviders:
  - name: github
    mappingMethod: claim
    type: GitHub
    github:
      clientID: "{{ github_clientid }}"
      clientSecret:
        name: github-oauth-app
      organizations:
      - BCDevOps
```

**Comment by StevenBarre** [comment]
I think the RBAC rules for OCP4 should be moved into the platform-ops repo with the code for the cron/operator.

Pros/Cons of cronjob vs operator?

**Comment by j-pye** [comment]
The other docs you pointed to are what I was hoping would work. The OCP4 docs weren't quite as clear. :) 

RBAC rules in the platform-ops repo makes sense to me.

TL;DR: The operator is probably faster and easier to implement but will use CRDs/CRs so current RBAC yaml files would need to be altered.

This is where I'm at: (Mid-thought)

**Cron - Pros:**
- Periodic resource usage
- Current RBAC yaml files can be used (untested)

**Cron - Cons:**  
- K8s cron jobs can be inconsistent when only a single job can run at a time  -- random failures
- Custom Ansible Image
- Slower updates due to builds or repository pull (deploy key)

**Operator - Pros**
- Faster to get up and running
- Build Once unless CRD changes
- CR updates are fast and added to the queue as soon as the CR changes
- Opportunity to reduce the amount of code required to write RBAC rules

**Operator - Cons**
- Constant resource usage (likely low)
- WIP - CRs will likely require an altered format for RBAC entry but will look the same as they do now when applied. So long as it's documented I don't really see an issue with this.

**Overview:**
Cron:
- Run every X hours
- Trigger on PR Merge
- Getting updates...
-- Option A: Pull RBAC from branch on repo every cron run -- Will require a read-only deploy key for the repo.
-- Option B: Build and Replace Cron Image on PR Merge which uses only the RBAC files added to the image

Operator:
- Build once unless there are updates to the CRDs
- Reconcile / Periodic runs built in
- Getting updates...
-- Update CRs on PR Merge

**Comment by jefkel** [comment]
I'm not sure where we're going with this.  Is there a reason why we're jumping into creating an operator for a cluster initialization/creation script?  (seems like scope creep)

I don't believe we were looking at creating an operator at this time for ongoing maintenance of RBAC rules, but rather the initial "Day 1 cluster configuration" automation.  Once we have the automation scripts that can build a "standard OCP 4 cluster", then I'd be happy to start planning on some ongoing operational tools (out of scope for this ticket though)

Based on the initial work that DXC has done for the openshift 3.11 clusters, this could be completed quite quickly with simply re-using the existing automation, and perhaps updating to be run on multiple clusters as per discussion with @sbarre-esit in https://github.com/bcgov-c/platform-ops/pull/189.

**Comment by j-pye** [comment]
@jefkel It was mentioned to me that it'd be nice to have something that ensures each clusters RBAC always matches the code in the repo such as a cronjob which is where the operator suggestion came from. 

If we just want a playbook that applies the RBAC rules that's fine but a task should be created to automate the maintenance of RBAC whatever solution that is. (I can't do that, i don't have enough permissions) 

I wasn't expecting either option to take a whole lot of time to implement and test for basic functionality but I'll go with just the basic playbook for now.

**Comment by jefkel** [comment]
I think an automated solution to ensure the RBAC matches the code will also require more work than you think (ie: what do you need to remove, not just what do you want to add...)  And as I said, I'm not against it, it's just not part of this scope. 

Once we get multiple clusters re-deploying based off our automated post-install configuration I'd be happy to have a conversation along the lines of https://cloud.google.com/anthos/config-management and what can we do in a similar way....

**Comment by stewartshea** [comment]
Folks, you shouldn't write this yourself... Think Flux or ArgoCD.
Eventually I'd like to take us down the Argo road for application
deployment; it might work just as well for cluster customizations as well,
such as RBAC manifests.
--
Shea Stewart, Partner
Arctiq: Intelligent Architecture
shea.stewart@arctiq.ca
+1-647-972-5191
http://www.arctiq.ca
Book at meeting here: https://x.ai/calendar/sheastewart

**Comment by StevenBarre** [comment]
I'd rather avoid two separate ways to do RBAC, one for "day one" and a separate one for ongoing management, I'd rather just deploy the ongoing tool on day one.

**Comment by filiy** [comment]
@jefkel @j-pye @sbarre-esit 

Some of this is automated already.

https://github.com/BCDevOps/OpenShift4-RollOut/blob/openshift4-azure/openshift4-azure/auto-config/azure-ocp4-postconfig.yml#L60

- [x] Set up GitHub authentication
- [x] Restricted project/namespace creation

Also add in the default project template (which would need an update for ocp4). request from SheaP

https://github.com/BCDevOps/OpenShift4-RollOut/blob/openshift4-azure/openshift4-azure/auto-config/azure-ocp4-postconfig.yml#L96

**Comment by filiy** [comment]
Also @j-pye, just give me a shout if you need a cluster in Azure to test. I can get one up today

**Comment by stewartshea** [comment]
I think you're missing the point on the larger conversation... how are you
going to centralize this across multiple clusters. We need to think GitOps
for multiple clusters and not single cluster configurations.
This also includes more checklist items than the above... there will be a
number of RBAC bindings that need to be deployed outside of  the
authorization provider (GitHub (which should pivot to SSO soon).
We will also want to easily roll in SSO when that's ready in OCP4.4 and
move away from direct GitHub, so that needs to be considered as well.

#discuss.
--
Shea Stewart, Partner
Arctiq: Intelligent Architecture
shea.stewart@arctiq.ca
+1-647-972-5191
http://www.arctiq.ca
Book at meeting here: https://x.ai/calendar/sheastewart

**Comment by StevenBarre** [comment]
MVP for now includes:

- A reduced set of privs from https://github.com/bcgov-c/devops-platform-operations-docs/tree/master/rbac with some basic automation. Will also set the project provisioning perms
- A reduced set of quotas from https://github.com/bcgov-c/platform-ops/tree/master/playbooks/run/quota-applyer with some basic automation. This also sets the default project template.

**Comment by StevenBarre** [comment]
@jefkel any progress on thoughts around how to separate config for clusters? Separate repos? forks? branches? folders?

**Comment by StevenBarre** [comment]
@j-pye said he has some half done code I can use as a jumping off point. Waiting on a PR to hand over that code so I know what else is remaining.

**Comment by jefkel** [comment]
I would lean toward separate repos with fork/pr's from a central point mainly because it could reduce the false-positive "changed" trigger for any tool (argo) is using to monitor it's config, while helping my brain keep the configuration repositories separated for each cluster.

That said, let's test out the operational process with the 2 cluster's we have and see which gitflow option you like better @sbarre-esit

**Comment by StevenBarre** [comment]
>  let's test out the operational process

Ugh, I dislike collective nouns in sentences. Everyone assumes someone else will do it :stuck_out_tongue:

**Comment by jefkel** [comment]
True, let me be more clear:

I would like @sbarre-esit and DXC team to test an operational process with the 2 clusters we have, and form an opinion on which gitflow they like better.  ;)
